### PR TITLE
[pipeline/e2e] end-to-end-tests.ps1 should fail with failure test

### PR DESCRIPTION
**Pull request by WangWeiLin-MV**
Fix `end-to-end-tests.ps1`, which should fail and stop with failure test case.

### Change

* Script end-to-end-tests.ps1
  + Check test `$LASTEXITCODE`
* Script end-to-end-tests-prelude.ps1
  + Reset `$LASTEXITCODE` for `Throw-IfNotFailed`
* Cases
  + autocomplete.ps1
    - Rename to `autocomplete-command.ps1`
  + env-tools.ps1
    - Add `Throw-IfNotFailed` for non-windows
  + ci-verify-versions.ps1
    - `Remove-Problem-Matchers` for malformed port
* Pipeline
  + Update to `action/checkout@4`, `actions/setup-node@v4`, `lukka/get-cmake@v3.31.0`
    - https://github.blog/changelog/2024-03-07-github-actions-all-actions-will-run-on-node20-instead-of-node16-by-default/

### CODE REVIEW

**Comment by WangWeiLin-MV** [comment]
### Test

* Example of un-expected pass before fixed
  + https://github.com/microsoft/vcpkg-tool/actions/runs/11435724127/job/31811675237?pr=1518#annotation:14:22
* Failed as expected
  + https://github.com/microsoft/vcpkg-tool/actions/runs/11436226023/job/31813205981?pr=1518#annotation:14:45
* Passed as expected
  + https://github.com/microsoft/vcpkg-tool/actions/runs/11436488593/job/31814031787?pr=1518

**Comment by BillyONeal** [comment]
>Fix end-to-end-tests.ps1, which should fail and stop with failure test case.

As far as I can tell it already does this. Do you have a repro for it not doing so?

**Comment by BillyONeal** [comment]
It looks like the repro might only happen with something like this "Pester" module ( as in https://github.com/microsoft/vcpkg-tool/pull/1512/ ) because all other existing e2e tests fail by throwing an exception, thus stopping further test runs.

**Comment by WangWeiLin-MV** [comment]
> > Fix end-to-end-tests.ps1, which should fail and stop with failure test case.
> 
> As far as I can tell it already does this. Do you have a repro for it not doing so?
> 
> (EDIT: I see you have an example in a comment here but it's not clear how you got that to happen)

Sorry I forgot to attach the test commit https://github.com/microsoft/vcpkg-tool/commit/77554801dcc44a7618832d2ee28ec5c4751cd33f

Run this script with `$ErrorActionPreference = "Stop"`, will get status `$?` to `False` and `$LASTEXITCODE` to `2`.

```powershell
BeforeAll {
    Import-Module $PSScriptRoot/not-exist/posh-vcpkg
}
Describe 'Module posh-vcpkg tests' {
    It 'Expect fail' {}
}
```

This issue has been fixed by:

* Check `$lastTextExitCode = $LASTEXITCODE`
* Reset `$LASTEXITCODE` for `Throw-IfNotFailed`

**Comment by WangWeiLin-MV** [comment]
I'm now trying to fix the error message `Unexpected character; expected property name` in GitHub action page, for example https://github.com/microsoft/vcpkg-tool/actions/runs/11450810448

**Comment by WangWeiLin-MV** [comment]
### Not changed

Unable change runner machine `runs-on` from `macos-12` to `macos-14`, even there aren't any build `macos-12` string in the repository.

It hangs with PR check:

```
builds / build (macos-12, macos-ci) Expected — Waiting for status to be reported
```

**Comment by WangWeiLin-MV** [comment]
### Note

`vmImage: macOS-12` is deprecated in `azure-pipelines/signing.yml:208`

https://devblogs.microsoft.com/devops/upcoming-deprecation-of-macos-12-hosted-pipeline-image/

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -80,6 +80,7 @@ jobs:
         shell: pwsh
         run: |
           cd out/build/${{ matrix.preset }}
+          Import-Module Pester -Force -MinimumVersion '5.6.1' -MaximumVersion '5.99'
```
Comment (BillyONeal) [inline .github/workflows/build.yaml:83]: Can you explain what this module is doing?

File: azure-pipelines/end-to-end-tests.ps1 [hunk 2]
```diff
@@ -158,7 +162,13 @@ for ($n = 1; $n -le $allTestsCount; $n++)
     if ($env:GITHUB_ACTIONS) {
         Write-Host "::endgroup::"
     }
+
+    if ($lastTestExitCode -ne 0)
+    {
+        Write-Host -ForegroundColor Red "[end-to-end-tests.ps1] Suite $testDisplayName failed with exit code $lastTestExitCode"
+        Write-Error 'Test failed.'
```
Comment (BillyONeal) [inline azure-pipelines/end-to-end-tests.ps1:169]: Why is this split into two writes?

File: .github/workflows/build.yaml (hunk 1 above)
Comment (WangWeiLin-MV) [reply .github/workflows/build.yaml:83]: Import Pester for #1512, or I should move the change to that PR?

File: azure-pipelines/end-to-end-tests.ps1 (hunk 2 above)
Comment (WangWeiLin-MV) [reply azure-pipelines/end-to-end-tests.ps1:169]: Changed to write once.

File: .github/workflows/build.yaml (hunk 1 above)
Comment (WangWeiLin-MV) [reply .github/workflows/build.yaml:83]: Moved to PR1512

**Comment by WangWeiLin-MV** [comment]
### Test

Expected fail with failed test case https://github.com/microsoft/vcpkg-tool/actions/runs/11925979891

**Comment by BillyONeal** [comment]
Thanks!

**Pull request merged**